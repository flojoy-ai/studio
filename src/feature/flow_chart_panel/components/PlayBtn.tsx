import { Button } from "@src/components/ui/button";
import useKeyboardShortcut from "@src/hooks/useKeyboardShortcut";
import { useFlowChartGraph } from "@src/hooks/useFlowChartGraph";
import { Node, Edge } from "reactflow";
import { ElementsData } from "flojoy/types";
import { Play } from "lucide-react";

interface PlayBtnProps {
  onPlay: (nodes: Node<ElementsData>[], edges: Edge[]) => void;
}

const PlayBtn = ({ onPlay }: PlayBtnProps) => {
  const { nodes, edges } = useFlowChartGraph();

  const handleClick = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.preventDefault();
    if (onPlay) {
      onPlay(nodes, edges);
    }
  };

  useKeyboardShortcut("ctrl", "p", () => onPlay(nodes, edges));
  useKeyboardShortcut("meta", "p", () => onPlay(nodes, edges));

  return (
    <Button size="sm" variant="outline" onClick={handleClick}>
      <Play size="20" />
      <div className="px-1"></div>
      <div>Play</div>
    </Button>
  );
};

export default PlayBtn;
