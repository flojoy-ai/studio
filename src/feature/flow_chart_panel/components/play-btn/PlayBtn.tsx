import { Button, createStyles, useMantineTheme } from "@mantine/core";
import React, { useRef } from "react";
import "@src/feature/flow_chart_panel/components/play-btn/play-btn.css";

interface PlayBtnProps {
  onClick?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  style?: React.CSSProperties;
  disabled?: boolean;
}

const useStyles = createStyles((theme) => {
  const accent =
    theme.colorScheme === "dark" ? theme.colors.accent1 : theme.colors.accent2;
  return {
    btnPlay: {
      width: 64,
      height: 32,
      fontSize: 12,
      fontWeight: 700,
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      gap: 6,
      flexWrap: "wrap",
      outline: 0,
      cursor: "pointer",
      position: "relative",
      transition: "transform ease-in 0.1s, box-shadow ease-in 0.25s",
      WebkitAppearance: "none",
      appearance: "none",
      backgroundColor: theme.colors.accent4[1],
      color: accent[0],
      border: `1px solid ${accent[0]}`,
      borderRadius: 6,
      "&:hover": {
        backgroundColor: accent[1] + "36",
      },
      "&:disabled, &:disabled:hover": {
        color: "#b5b5b5",
        backgroundColor: "#e9ecef",
        border: 0,
        cursor: "not-allowed",
      },
      "&:disabled > svg > path": {
        fill: "#b5b5b5",
      },
      "&:focus": {
        outline: 0,
      },
    },
  };
});

const PlayBtn = ({ onClick, style, disabled = false }: PlayBtnProps) => {
  const theme = useMantineTheme();
  const { classes } = useStyles();

  const ButtonElem = useRef<HTMLButtonElement>(null);
  const handleClick = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.preventDefault();
    if (ButtonElem.current) {
      ButtonElem.current.classList.add("animate");
      //reset animation
      setTimeout(() => {
        ButtonElem.current?.classList.remove("animate");
      }, 1000);
    }
    if (onClick) {
      onClick(e);
    }
  };

  return (
    <Button
      className={classes.btnPlay}
      ref={ButtonElem}
      style={style}
      onClick={handleClick}
      data-cy="btn-play"
      data-testid="btn-play"
      disabled={disabled}
      title={disabled ? "Server is offline" : "Run Script"}
    >
      <svg
        width="9"
        height="11"
        viewBox="0 0 9 11"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
        style={{ marginBottom: 2 }}
      >
        <path
          d="M8.5 4.63397C9.16667 5.01887 9.16667 5.98113 8.5 6.36603L1.75 10.2631C1.08333 10.648 0.25 10.1669 0.25 9.39711L0.25 1.60289C0.25 0.833085 1.08333 0.35196 1.75 0.73686L8.5 4.63397Z"
          fill={
            theme.colorScheme === "light"
              ? theme.colors.accent2[0]
              : theme.colors.accent1[0]
          }
        />
      </svg>
      <div style={{ padding: "4px" }} />
      <span>Play</span>
    </Button>
  );
};

export default PlayBtn;
