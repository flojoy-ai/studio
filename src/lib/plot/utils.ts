export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
