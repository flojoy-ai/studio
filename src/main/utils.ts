import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
