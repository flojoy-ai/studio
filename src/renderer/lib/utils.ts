import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
