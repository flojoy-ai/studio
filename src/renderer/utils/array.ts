export function filterMap<T, U>(
  arr: Array<T>,
  mapPred: (x: T) => U | undefined | null,
) {
  return arr.reduce((acc: Array<U>, x: T) => {
    const val = mapPred(x);
    if (val !== undefined && val !== null) {
      acc.push(val);
    }

    return acc;
  }, []);
}
