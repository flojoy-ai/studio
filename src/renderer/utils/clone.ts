import { DeepMutable } from "@/renderer/types/util";

export const deepMutableClone = <T>(obj: T): DeepMutable<T> => {
  return JSON.parse(JSON.stringify(obj)) as DeepMutable<T>;
};
