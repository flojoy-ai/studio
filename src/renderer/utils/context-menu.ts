export const calculateContextMenuOffset = (
  clickX: number,
  clickY: number,
  flow: HTMLDivElement,
) => {
  // Calculate position of the context menu. We want to make sure it
  // doesn't get positioned off-screen.
  const pane = flow.getBoundingClientRect();

  const contextMenuHeight = 200;
  const paneToBlock = clickY - contextMenuHeight;

  let top: number | undefined = undefined;
  let bottom: number | undefined = undefined;

  if (paneToBlock < contextMenuHeight / 2) {
    top = paneToBlock;
  } else if (paneToBlock < contextMenuHeight) {
    if (pane.height - paneToBlock < contextMenuHeight) {
      top = contextMenuHeight - paneToBlock;
    } else {
      top = paneToBlock;
    }
  } else if (pane.height - paneToBlock < contextMenuHeight) {
    top = undefined;
    bottom = pane.height - paneToBlock;
  } else {
    top = paneToBlock;
  }

  return {
    top,
    left: clickX < pane.width - 200 ? clickX : undefined,
    right: clickX >= pane.width - 200 ? pane.width - clickX : undefined,
    bottom,
  };
};
