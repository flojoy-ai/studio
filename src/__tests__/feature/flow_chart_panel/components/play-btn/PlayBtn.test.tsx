import { fireEvent, screen } from "@testing-library/react";
import PlayBtn from "@src/feature/flow_chart_panel/components/play-btn/PlayBtn";
import { renderWithTheme } from "@src/__tests__/__utils__/utils";

describe("PlayBtn component", () => {
  it("renders correctly with default props", () => {
    renderWithTheme(<PlayBtn onPlay={jest.fn()} />);
    const playBtn = screen.getByRole("button");
    expect(playBtn).toBeInTheDocument();
    expect(playBtn).not.toBeDisabled();
    expect(playBtn).toHaveAttribute("title", "Run Script");
    expect(playBtn).toMatchSnapshot();
  });

  it("calls onClick function when clicked", () => {
    const handleClick = jest.fn();
    renderWithTheme(<PlayBtn onPlay={handleClick} />);
    const playBtn = screen.getByRole("button");
    fireEvent.click(playBtn);
    expect(handleClick).toHaveBeenCalled();
  });

  it("adds animation class on click", () => {
    renderWithTheme(<PlayBtn onPlay={jest.fn()} />);
    const playBtn = screen.getByRole("button");
    fireEvent.click(playBtn);
    expect(playBtn).toHaveClass("animate");
  });

  it("removes animation class after 1 second", async () => {
    jest.useFakeTimers();
    renderWithTheme(<PlayBtn onPlay={jest.fn()} />);
    const playBtn = screen.getByRole("button");
    fireEvent.click(playBtn);
    expect(playBtn).toHaveClass("animate");
    jest.advanceTimersByTime(1000);
    expect(playBtn).not.toHaveClass("animate");
  });

  it("disables button when 'disabled' prop is true", () => {
    renderWithTheme(<PlayBtn onPlay={jest.fn()} disabled={true} />);
    const playBtn = screen.getByRole("button");
    expect(playBtn).toBeDisabled();
    expect(playBtn).toHaveAttribute("title", "Server is offline");
  });
});
