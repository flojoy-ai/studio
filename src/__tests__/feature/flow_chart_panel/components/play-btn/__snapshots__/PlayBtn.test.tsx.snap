// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayBtn component renders correctly with default props 1`] = `
<button
  class="mantine-UnstyledButton-root mantine-Button-root mantine-btd47x"
  data-button="true"
  data-cy="btn-play"
  data-testid="btn-play"
  title="Run Script"
  type="button"
>
  <div
    class="mantine-1wpc1xj mantine-Button-inner"
  >
    <span
      class="mantine-1ryt1ht mantine-Button-label"
    >
      <div>
        <svg
          fill="none"
          height="11"
          style="margin-bottom: 2px;"
          viewBox="0 0 9 11"
          width="9"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.5 4.63397C9.16667 5.01887 9.16667 5.98113 8.5 6.36603L1.75 10.2631C1.08333 10.648 0.25 10.1669 0.25 9.39711L0.25 1.60289C0.25 0.833085 1.08333 0.35196 1.75 0.73686L8.5 4.63397Z"
            fill="#99f5ff"
          />
        </svg>
      </div>
      <div
        style="padding: 4px;"
      />
      <span>
        Play
      </span>
    </span>
  </div>
</button>
`;
